src/batSet.ml

```diff
@@ -531,6 +679,28 @@ module Concrete = struct
         subset cmp (Node (l1, v1, Empty, 0)) l2 && subset cmp r1 t2
       else
         subset cmp (Node (Empty, v1, r1, 0)) r2 && subset cmp l1 t2
+
+  let add_seq cmp s m =
+    Seq.fold_left (fun m e -> add cmp e m) m s
+    
+  let of_seq cmp s =
+    add_seq cmp s empty
+    
+  let rec to_seq m =
+    match m with
+    | Empty -> Seq.empty
+    | Node(l, v, r, _) -> 
+       Seq.append (to_seq l) (Seq.cons v (to_seq r))
```

Comment:
Two remarks:
- This implementation does not do what I would expect due to the use of the strict function `Seq.cons`: it will traverse the whole set before returning the first element. This is not technically wrong as there is no order-of-evaluation difference (and probably not much of a performance difference, although it is worth measuring), but I would find it more natural to have a version that lazily enumerates the list.
- We have a proper datatype to represent partial enumerations of sets, `'a iter` and `cons_iter`.

I would expect something looking roughly like (untested):
```
let rec iter_to_seq = function
| E -> Seq.empty
| C (e, r, t) -> fun () -> Seq.Cons (e, iter_to_seq (cons_iter r t))

let to_seq s = iter_to_seq (cons_iter t E)
```

(`to_seq_from` also needs an adaptation.)

I would be curious to know what is the performance difference between your version, a version using `iter` as above, and also a modified version of your code that avoids `Seq.cons` to avoid forcing the tail (but otherwise traverses the set recursively in the direct way).

Reply:
I modified to_seq and to_seq_from to be more lazy.
They will not do anything on invocation (except returning a closure).
Furthermore, once that closure is called, it will only go down the leftmost path in the tree to determine the first element; recursing into the right subtrees is again done in a closure.

I hope that satisfies your expectations of these functions.

Reply:
They do, thanks! I would still be curious to know whether the `'a iter` version is faster, but I don't have the time to look at this right now and maybe you are not so curious yourself, so the new versions are fine with me.